extensions/oplogPopulator/modules/Connector.js

```diff
@@ -253,11 +273,15 @@ class Connector {
      * Connector is updated with new bucket list
      * @param {string} bucket bucket to add
      * @param {boolean} [doUpdate=false] updates connector if true
-     * @returns {Promise|undefined} undefined
+     * @returns {Promise<boolean>} true if bucket was removed from
+     * the bucket
      * @throws {InternalError}
      */
     async removeBucket(bucket, doUpdate = false) {
         this._buckets.delete(bucket);
+        // Here, we must avoid removing the bucket from the
+        // bucketsInUse set, as the connector might still be
+        // using it in a deployed pipeline.
```

Comment:
Here is my attempt after our discussion: https://github.com/scality/backbeat/pull/2539/commits/d4bce59e27e11458a178160a5b7f92e9fde8d670

-> I basically moved the responsibility to "allocate" to the strategies, including the map of bucket-> connectors, notification from the connectorsManager that the pipelines are updated to clear the reinated buckets, and handling of the logic to listen/stop listen to a bucket. This way, the Allocator is very simple and wan work with any strategy. the strategies are now a bit more complex, but it is very linear... Please let me know what you think.

Reply:
Too much duplication is a sign that we have the bad level of abstraction...

With your current approach this can be easily fixed by having 2 methods, 1 "generic" in `AllocationStrategy` (which does the allocation) and 1 "simple" the actual strategy implementations.... However this means the AllocationStrategy is not an interface anymore, but really an Allocator...

C.f. another comment on how to deal with the retained bucket "in isolation". And looking at the code I did not see a place where the strategies needed the bucket->connector mapping.